Add "function" to the glossary plus related improvements (#11144)
crate/crate: docs/sql/statements/select.rst

  is combined into groups of rows that match on one or more values.

- The actual output rows are computed using the ``SELECT`` output expressions
  for each selected row or row group.

- If the :ref:`ORDER BY <sql-select-order-by>` clause is specified, the
  returned rows are sorted in the specified order. If ``ORDER BY`` is not
  given, the rows are returned in whatever order the system finds fastest to
  produce.

- If :ref:`DISTINCT <sql-select-list>` is specified, one unique row is
  kept. All other duplicate rows are removed from the result set.

- If the :ref:`LIMIT <sql-select-limit>` or :ref:`OFFSET <sql-select-offset>`
  clause is specified, the ``SELECT`` statement only returns a subset of the
  result rows.

.. _sql-select-parameters:

Parameters
==========

.. _sql-select-list:

The ``SELECT`` List
-------------------

The ``SELECT`` list specifies expressions that form the output rows of the
``SELECT`` statement. The expressions can (and usually do) refer to columns
computed in the ``FROM`` clause.

::

    SELECT [ ALL | DISTINCT ] * | expression [ [ AS ] output_name ] [, ...]

Just as in a table, every output column of a ``SELECT`` has a name. In a simple

> ``SELECT`` this name is just used to label the column for display. To specify

Reviewer: ```suggestion
``SELECT``, this name is just used to label the column for display. To specify
```

the name to use for an output column, write ``AS output_name`` after the
column's ``expression``. (You can omit ``AS``, but only if the desired output
name does not match any reserved keyword. For protection against possible
future keyword additions, it is recommended that you always either write ``AS``
or double-quote the output name.) If you do not specify a column name, a name
is chosen automatically by CrateDB. If the column's expression is a simple

> column reference then the chosen name is the same as that column's name. In

Reviewer: ```suggestion
column reference, then the chosen name is the same as that column's name. In
```

> more complex cases a :ref:`function <gloss-function>` or type name may be used,

Reviewer: ```suggestion
more complex cases, a :ref:`function <gloss-function>` or type name may be used,
```

  A where condition is any expression that evaluates to a result of type
  boolean.

  Any row that does not satisfy this condition will be eliminated from the
  output. A row satisfies the condition if it returns true when the actual row
  values are substituted for any variable references.

.. _sql-select-group-by:

``GROUP BY``
............

> The optional ``GROUP BY`` clause will condense into a single row all selected
> rows that share the same values for the grouped expressions.

Reviewer: ```suggestion
The optional ``GROUP BY`` clause will condense all selected rows that
share the same values for the grouped expressions into a single row.
```

:ref:`Aggregate expressions <aggregation-expressions>`, if any are used, are
computed across all rows making up each group, producing a separate value for
each group.

::

    GROUP BY expression [, ...] [HAVING condition]

:expression:
  An arbitrary :ref:`expression <sql-value-expressions>` formed from column
  references of the queried relation that are also present in the result column
  list. Numeric literals are interpreted as ordinals referencing an output
  column from the select list.

  It can also reference output columns by name.

  In case of ambiguity, a ``GROUP BY`` name will be interpreted as a name of a
  column from the queried relation rather than an output column name.

.. _sql-select-having:

``HAVING``
''''''''''

The optional ``HAVING`` clause defines the condition to be met for values

> within a resulting row of a group by clause.

Reviewer: ```suggestion
within a resulting row of a ``GROUP BY`` clause.
```

:condition:

>   A having condition is any expression that evaluates to a result of type

Reviewer: ```suggestion
  A ``HAVING`` condition is any expression that evaluates to a result of type
```

.. _sql-select-union-all:

UNION ALL
.........

The ``UNION ALL`` :ref:`operator <gloss-operator>` combines the result sets of
two or more ``SELECT`` statements. The two ``SELECT`` statements that represent
the direct :ref:`operands <gloss-operand>` of the ``UNION ALL`` must produce
the same number of columns, and corresponding columns must be of the same data
types. The result of ``UNION ALL`` may contain duplicate rows. You can find
:ref:`here <sql_union>` sample usage of ``UNION ALL``.

::

    UNION ALL query_specification

:query_specification:
  Can be any ``SELECT`` statement.

``ORDER BY``, ``LIMIT``, and ``OFFSET`` can only be applied after the last
``SELECT`` statement of the ``UNION ALL``, as they are applied to the complete
result of the ``UNION`` operation. In order to apply an ``ORDER BY`` and/or
``LIMIT`` and/or ``OFFSET`` to any of the partial ``SELECT`` statements, those
statements need to become subqueries.

Column names used in ``ORDER BY`` must be position numbers or refer to the
outputs of the first ``SELECT`` statement, and no :ref:`functions
<gloss-function>` can be applied on top of the ``ORDER BY`` symbols. To achieve

> more complex ordering ``UNION ALL`` must become a subselect and the more

Reviewer: ```suggestion
more complex ordering, ``UNION ALL`` must become a subselect and the more
```

complex ``ORDER BY`` should be applied on the outer ``SELECT`` wrapping the
``UNION ALL`` subselect.

> Ordering of the outcome is not guaranteed unless ``ORDER BY`` is used.

Reviewer: ```suggestion
The ordering of the outcome is not guaranteed unless ``ORDER BY`` is used.
```

  Can be the name or ordinal number of an output column, or it can be an
  arbitrary expression formed from input-column values.

The optional keyword ``ASC`` (ascending) or ``DESC`` (descending) after any

> expression allows to define the direction in which values have are sorted. The

Reviewer: ```suggestion
expression allows to define the direction in which values are sorted. The
```

default is ascending.

> If ``NULLS FIRST`` is specified, null values sort before non null values. If
> ``NULLS LAST`` is specified null values sort after non null values.  If neither
> is specified nulls are considered larger than any value. That means the default
> for ``ASC`` is ``NULLS LAST`` and the default for ``DESC`` is ``NULLS FIRST``.

Reviewer: ```suggestion
If ``NULLS FIRST`` is specified, null values sort before non-null values. If
``NULLS LAST`` is specified, null values sort after non-null values.  If neither
is specified, nulls are considered larger than any value. That means the default
for ``ASC`` is ``NULLS LAST`` and the default for ``DESC`` is ``NULLS FIRST``.
```

.. NOTE::

    Sorting can only be applied on indexed fields.

    Additionally, sorting on :ref:`geo_point_data_type`,
    :ref:`geo_shape_data_type`, :ref:`data-type-array`, and
    :ref:`object_data_type` is not supported.

.. SEEALSO::

    :ref:`Fulltext indices : Disable indexing <sql_ddl_index_off>`

.. _sql-select-window:

``WINDOW``
..........

The optional ``WINDOW`` clause has a form:

::

   WINDOW window_name AS ( window_definition ) [, ...]

The ``window_name`` is a name that can be referenced from ``OVER`` clauses or
subsequent window definitions.

The ``window_definition`` determines the partitioning and ordering of rows
before the :ref:`window function <window-functions>` is applied.

.. SEEALSO::

    :ref:`Window functions: Window definition <window-definition>`

    :ref:`Window functions: Named windows <window-definition-named-windows>`

.. _sql-select-limit:

``LIMIT``
.........

> The optional ``LIMIT`` clause allows to limit the number or retured result

Reviewer: ```suggestion
The optional ``LIMIT`` clause allows to limit the number of returned result
```

.. NOTE::

    Sorting can only be applied on indexed fields.

    Additionally, sorting on :ref:`geo_point_data_type`,
    :ref:`geo_shape_data_type`, :ref:`data-type-array`, and
    :ref:`object_data_type` is not supported.

.. SEEALSO::

    :ref:`Fulltext indices : Disable indexing <sql_ddl_index_off>`

.. _sql-select-window:

``WINDOW``
..........

The optional ``WINDOW`` clause has a form:

::

   WINDOW window_name AS ( window_definition ) [, ...]

The ``window_name`` is a name that can be referenced from ``OVER`` clauses or
subsequent window definitions.

The ``window_definition`` determines the partitioning and ordering of rows
before the :ref:`window function <window-functions>` is applied.

.. SEEALSO::

    :ref:`Window functions: Window definition <window-definition>`

    :ref:`Window functions: Named windows <window-definition-named-windows>`

.. _sql-select-limit:

``LIMIT``
.........

> The optional ``LIMIT`` clause allows to limit the number of retured result

Reviewer: ```suggestion
The optional ``LIMIT`` clause allows to limit the number of returned result
```
